File: src/main/java/com/adven/concordion/extensions/exam/db/commands/DBCheckCommand.java
package com.adven.concordion.extensions.exam.db.commands;

import com.adven.concordion.extensions.exam.db.DbResultRenderer;
import com.adven.concordion.extensions.exam.html.Html;
import org.concordion.api.CommandCall;
import org.concordion.api.Element;
import org.concordion.api.Evaluator;
import org.concordion.api.ResultRecorder;
import org.concordion.api.listener.AssertEqualsListener;
import org.concordion.api.listener.AssertFailureEvent;
import org.concordion.api.listener.AssertSuccessEvent;
import org.concordion.internal.util.Announcer;
import org.dbunit.DatabaseUnitException;
import org.dbunit.IDatabaseTester;
import org.dbunit.assertion.DbComparisonFailure;
import org.dbunit.assertion.DiffCollectingFailureHandler;
import org.dbunit.assertion.Difference;
import org.dbunit.database.IDatabaseConnection;
import org.dbunit.dataset.Column;
import org.dbunit.dataset.DataSetException;
import org.dbunit.dataset.ITable;
import org.dbunit.dataset.SortedTable;
import org.dbunit.util.QualifiedTableName;

import java.util.List;

import static com.adven.concordion.extensions.exam.html.Html.*;
import static com.google.common.base.Strings.isNullOrEmpty;
import static org.concordion.api.Result.FAILURE;
import static org.concordion.api.Result.SUCCESS;
import static org.dbunit.Assertion.assertEquals;
import static org.dbunit.dataset.filter.DefaultColumnFilter.includedColumnsTable;

public class DBCheckCommand extends DBCommand {
    private Announcer<AssertEqualsListener> listeners = Announcer.to(AssertEqualsListener.class);

    public DBCheckCommand(String name, String tag, IDatabaseTester dbTester) {
        super(name, tag, dbTester);
        listeners.addListener(new DbResultRenderer());
    }

    private void failure(ResultRecorder resultRecorder, Element element, Object actual, String expected) {
        resultRecorder.record(FAILURE);
        listeners.announce().failureReported(new AssertFailureEvent(element, expected, actual));
    }

    private void success(ResultRecorder resultRecorder, Element element) {
        resultRecorder.record(SUCCESS);
        listeners.announce().successReported(new AssertSuccessEvent(element));
    }

    @Override
    public void verify(CommandCall commandCall, Evaluator evaluator, ResultRecorder resultRecorder) {
        try {
            final ITable actual = getActualTable();
            final ITable filteredActual = includedColumnsTable(actual, getExpectedTable().getTableMetaData().getColumns());

            assertEq(new Html(commandCall.getElement()), resultRecorder, getExpectedTable(), filteredActual);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private ITable getActualTable() throws Exception {
        IDatabaseConnection conn = getDbTester().getConnection();
        String tableName = getExpectedTable().getTableMetaData().getTableName();
        String qualifiedName = new QualifiedTableName(tableName, conn.getSchema()).getQualifiedName();

        return conn.createQueryTable(qualifiedName, "select * from " + qualifiedName + where());
    }

    private String where() {
        return isNullOrEmpty(getWhere()) ? "" : " WHERE " + getWhere();
    }

    private void assertEq(Html root, ResultRecorder resultRecorder, ITable expected, ITable actual)
            throws DatabaseUnitException {
        final DiffCollectingFailureHandler diffHandler = new DiffCollectingFailureHandler();
        Column[] columns = expected.getTableMetaData().getColumns();
        SortedTable expectedTable = new SortedTable(expected, columns);
        try {
            assertEquals(expectedTable, new SortedTable(actual, columns), diffHandler);
        } catch (DbComparisonFailure f) {
            //TODO move to ResultRenderer
            resultRecorder.record(FAILURE);
            Html div = div().css("rest-failure bd-callout bd-callout-danger").childs(div(f.getMessage()));
            root.below(div);

            Html exp = tableSlim();
            div.childs(span("Expected: "), exp);
            root = exp;

            Html act = Html.tableSlim();
            renderTable(act, actual);
            div.childs(span("but was: "), act);
        }
        for (Difference diff : (List<Difference>) diffHandler.getDiffList()) {
            System.err.println("***** DIFF " + diff.toString());
        }
        checkResult(root, expectedTable, diffHandler.getDiffList(), resultRecorder);
    }

    private void checkResult(Html el, ITable expected, List<Difference> diffs, ResultRecorder resultRecorder) {
        try {
            String title = el.attr("caption");
            el.childs(dbCaption(expected, title));

            Column[] cols = expected.getTableMetaData().getColumns();
            Html header = thead();
            Html thr = tr();
            for (Column col : cols) {
                thr.childs(th(col.getColumnName()));
            }
            el.childs(header.childs(thr));

            if (expected.getRowCount() == 0) {
                Html td = td("<EMPTY>").attr("colspan", String.valueOf(cols.length));
                Html tr = tr().childs(td);
                el.childs(tr);
                success(resultRecorder, td.el());
            } else {
                for (int i = 0; i < expected.getRowCount(); i++) {
                    Html tr = tr();
                    for (Column col : cols) {
                        Object expectedValue = expected.getValue(i, col.getColumnName());
                        String displayedExpected = expectedValue == null ? "(null)" : expectedValue.toString();
                        Html td = td(displayedExpected);
                        tr.childs(td);
                        Difference fail = findFail(diffs, i, col);
                        if (fail != null) {
                            failure(resultRecorder, td.el(), fail.getActualValue(), displayedExpected);
                        } else {
                            success(resultRecorder, td.el());
                        }
                    }
                    el.childs(tr);
                }
            }
        } catch (DataSetException e) {
            throw new RuntimeException(e);
        }
    }

    private Difference findFail(List<Difference> diffs, int i, Column col) {
        Difference fail = null;
        for (Difference diff : diffs) {
            if (diff.getRowIndex() == i && diff.getColumnName().equals(col.getColumnName())) {
                fail = diff;
                break;
            }
        }
        return fail;
    }
}

File: src/main/java/com/adven/concordion/extensions/exam/db/commands/DBCheckCommand.kt
package com.adven.concordion.extensions.exam.db.commands

import com.adven.concordion.extensions.exam.core.html.*
import com.adven.concordion.extensions.exam.core.resolveToObj
import com.adven.concordion.extensions.exam.core.utils.parsePeriod
import com.adven.concordion.extensions.exam.db.DbPlugin
import com.adven.concordion.extensions.exam.db.DbResultRenderer
import com.adven.concordion.extensions.exam.db.DbTester
import com.adven.concordion.extensions.exam.db.RowComparator
import org.awaitility.Awaitility
import org.awaitility.core.ConditionTimeoutException
import org.concordion.api.CommandCall
import org.concordion.api.Evaluator
import org.concordion.api.Result.FAILURE
import org.concordion.api.Result.SUCCESS
import org.concordion.api.ResultRecorder
import org.concordion.api.listener.AssertEqualsListener
import org.concordion.api.listener.AssertFailureEvent
import org.concordion.api.listener.AssertSuccessEvent
import org.concordion.internal.util.Announcer
import org.dbunit.Assertion
import org.dbunit.assertion.DbComparisonFailure
import org.dbunit.assertion.Difference
import org.dbunit.assertion.comparer.value.IsActualEqualToExpectedValueComparer
import org.dbunit.assertion.comparer.value.IsActualWithinToleranceOfExpectedTimestampValueComparer
import org.dbunit.dataset.CompositeTable
import org.dbunit.dataset.ITable
import org.dbunit.dataset.SortedTable
import org.dbunit.dataset.datatype.DataType
import org.dbunit.util.QualifiedTableName
import org.joda.time.LocalDateTime
import java.sql.Timestamp
import java.util.*
import java.util.concurrent.TimeUnit
import java.util.regex.Pattern

class DBCheckCommand(
    name: String,
    tag: String,
    dbTester: DbTester,
    valuePrinter: DbPlugin.ValuePrinter
) : DBCommand(name, tag, dbTester, valuePrinter) {
    private val listeners = Announcer.to(AssertEqualsListener::class.java)

    private val actualTable: ITable
        get() {
            val conn = dbTester.connectionFor(ds)
            val qualifiedName = QualifiedTableName(expectedTable.tableName(), conn.schema).qualifiedName
            val where = if (where.isNullOrEmpty()) "" else "WHERE $where"
            return conn.createQueryTable(qualifiedName, "select * from $qualifiedName $where")
        }

    init {
        listeners.addListener(DbResultRenderer())
    }

    private fun failure(resultRecorder: ResultRecorder, html: Html, actual: Any?, expected: String): Html {
        resultRecorder.record(FAILURE)
        listeners.announce().failureReported(AssertFailureEvent(html.el, expected, actual))
        return html
    }

    private fun success(resultRecorder: ResultRecorder, html: Html): Html {
        resultRecorder.record(SUCCESS)
        listeners.announce().successReported(AssertSuccessEvent(html.el))
        return html
    }

    override fun verify(cmd: CommandCall?, evaluator: Evaluator?, resultRecorder: ResultRecorder?) {
        dbTester.dbUnitConfig.valueComparer.setEvaluator(evaluator!!)
        dbTester.dbUnitConfig.columnValueComparers.forEach { it.value.setEvaluator(evaluator) }
        assertEq(cmd.html(), resultRecorder)
    }

    private fun assertEq(rootEl: Html, resultRecorder: ResultRecorder?) {
        var root = rootEl
        val sortCols: Array<String> = if (orderBy.isEmpty()) expectedTable.columnNamesArray() else orderBy
        var actual = sortedTable(actualTable.withColumnsAsIn(expectedTable), sortCols, dbTester.dbUnitConfig.overrideRowSortingComparer)
        val expected = sortedTable(CompositeTable(actual.tableMetaData, expectedTable), sortCols, dbTester.dbUnitConfig.overrideRowSortingComparer)
        val atMostSec = root.takeAwayAttr("awaitAtMostSec")
        val pollDelay = root.takeAwayAttr("awaitPollDelayMillis")
        val pollInterval = root.takeAwayAttr("awaitPollIntervalMillis")
        try {
            if (atMostSec != null || pollDelay != null || pollInterval != null) {
                val atMost = atMostSec?.toLong() ?: 4
                val delay = pollDelay?.toLong() ?: 0
                val interval = pollInterval?.toLong() ?: 1000
                try {
                    Awaitility.await("Await DB table ${expected.tableName()}")
                        .atMost(atMost, TimeUnit.SECONDS)
                        .pollDelay(delay, TimeUnit.MILLISECONDS)
                        .pollInterval(interval, TimeUnit.MILLISECONDS)
                        .untilAsserted {
                            actual = sortedTable(actualTable.withColumnsAsIn(expectedTable), sortCols, dbTester.dbUnitConfig.overrideRowSortingComparer)
                            dbUnitAssert(expected, actual)
                            if (dbTester.dbUnitConfig.diffFailureHandler.diffList.isNotEmpty()) {
                                throw AssertionError()
                            }
                        }
                } catch (f: ConditionTimeoutException) {
                    root(pre(
                        "DB check with poll delay $delay ms and poll interval $interval ms didn't complete within $atMost seconds:"
                    ).css("alert alert-danger small"))
                    if (f.cause is DbComparisonFailure) {
                        throw f.cause as DbComparisonFailure
                    }
                }
            } else {
                dbUnitAssert(expected, actual)
            }
        } catch (f: DbComparisonFailure) {
            root = rowsCountMismatch(resultRecorder, f, root, actual)
        } finally {
            checkResult(root, expected, actual, dbTester.dbUnitConfig.diffFailureHandler.diffList as List<Difference>, resultRecorder!!)
        }
    }

    //TODO move to ResultRenderer
    private fun rowsCountMismatch(resultRecorder: ResultRecorder?, f: DbComparisonFailure, root: Html, actual: SortedTable): Html {
        var root1 = root
        resultRecorder!!.record(FAILURE)
        val div = div().css("rest-failure bd-callout bd-callout-danger")(div(f.message))
        root1.below(div)

        val exp = tableSlim()
        div(span("Expected: "), exp)
        root1 = exp

        val act = tableSlim()
        renderTable(act, actual, remarks, valuePrinter)
        div(span("but was: "), act)
        return root1
    }

    private fun dbUnitAssert(expected: SortedTable, actual: ITable) {
        dbTester.dbUnitConfig.diffFailureHandler.diffList.clear()
        Assertion.assertWithValueComparer(
            expected,
            actual,
            dbTester.dbUnitConfig.diffFailureHandler,
            dbTester.dbUnitConfig.valueComparer,
            dbTester.dbUnitConfig.columnValueComparers
        )
    }

    private fun checkResult(root: Html, expected: ITable, actual: ITable, diffs: List<Difference>, resultRecorder: ResultRecorder) {
        val markAsSuccessOrFailure: (Html, Int, String) -> Html = { td, row, col ->
            val value = expected[row, col]
            val expectedValue = valuePrinter.wrap(value)
            diffs.firstOrNull { diff ->
                diff.rowIndex == row && diff.columnName == col
            }?.markAsFailure(resultRecorder, td)
                ?: td.markAsSuccess(resultRecorder)(
                    Html(expectedValue).text(
                        appendIf(isDbMatcher(value) && actual.rowCount == expected.rowCount, actual, row, col)
                    )
                )
        }
        renderTable(root, expected, markAsSuccessOrFailure, ifEmpty = { markAsSuccess(resultRecorder) })
    }

    private fun appendIf(append: Boolean, actual: ITable, row: Int, col: String): String = if (append) " (${actual[row, col]})" else ""

    private fun Html.markAsSuccess(resultRecorder: ResultRecorder) = success(resultRecorder, this)
    private fun Difference.markAsFailure(resultRecorder: ResultRecorder, td: Html): Html {
        return failure(resultRecorder, td, this.actualValue, valuePrinter.print(this.expectedValue))
    }

    companion object {
        fun isDbMatcher(text: Any?) =
            text is String && (text.isRegex() || text.isWithin() || text.isNumber() || text.isString() || text.isNotNull())
    }
}

class WithinValueComparer(tolerance: Long) : IsActualWithinToleranceOfExpectedTimestampValueComparer(0, tolerance) {
    public override fun isExpected(
        expectedTable: ITable?,
        actualTable: ITable?,
        rowNum: Int,
        columnName: String?,
        dataType: DataType,
        expectedValue: Any?,
        actualValue: Any?
    ) = super.isExpected(expectedTable, actualTable, rowNum, columnName, dataType, expectedValue, actualValue)

    override fun convertValueToTimeInMillis(timestampValue: Any?) = if (timestampValue is java.sql.Date) timestampValue.time
    else super.convertValueToTimeInMillis(timestampValue)
}

open class RegexAndWithinAwareValueComparer : IsActualEqualToExpectedValueComparer() {
    protected lateinit var evaluator: Evaluator

    fun setEvaluator(evaluator: Evaluator): RegexAndWithinAwareValueComparer {
        this.evaluator = evaluator
        return this
    }

    override fun isExpected(
        expectedTable: ITable?,
        actualTable: ITable?,
        rowNum: Int,
        columnName: String?,
        dataType: DataType,
        expected: Any?,
        actual: Any?
    ): Boolean = when {
        expected.isNotNull() -> setVarIfNeeded(actual, expected) { a, _ -> a != null }
        expected.isNumber() -> setVarIfNeeded(actual, expected) { a, _ -> regexMatches("^\\d+\$", a) }
        expected.isString() -> setVarIfNeeded(actual, expected) { a, _ -> regexMatches("^\\w+\$", a) }
        expected.isRegex() -> setVarIfNeeded(actual, expected) { a, e -> regexMatches(e, a) }
        expected.isWithin() -> setVarIfNeeded(actual, expected) { a, e ->
            WithinValueComparer(expected.toString().withinPeriod()).isExpected(
                expectedTable, actualTable, rowNum, columnName, dataType, resolve(e.toString()), a
            )
        }
        else -> super.isExpected(expectedTable, actualTable, rowNum, columnName, dataType, expected, actual)
    }

    private fun setVarIfNeeded(actual: Any?, expected: Any?, check: (actual: Any?, expected: Any?) -> Boolean): Boolean {
        val split = expected.toString().split(">>")
        if (split.size > 1) evaluator.setVariable("#${split[1]}", actual)
        return check(actual, split[0])
    }

    private fun resolve(expected: String): Timestamp {
        val expectedDateExpression = expected.substring(expected.indexOf("}") + 1).trim()
        return Timestamp(
            (if (expectedDateExpression.isBlank()) Date() else (evaluator.resolveToObj(expectedDateExpression) as Date))
                .time
        )
    }

    private fun regexMatches(expectedValue: Any?, actualValue: Any?): Boolean {
        if (actualValue == null) return false
        val expected = expectedValue.toString()
        return regexMatches(expected.substring(expected.indexOf("}") + 1).trim(), actualValue)
    }

    private fun regexMatches(pattern: String, actualValue: Any?): Boolean =
        if (actualValue == null) false else Pattern.compile(pattern).matcher(actualValue.toString()).matches()
}

class IgnoreMillisComparer : RegexAndWithinAwareValueComparer() {
    override fun isExpected(
        expectedTable: ITable?, actualTable: ITable?, rowNum: Int, columnName: String?, dataType: DataType, expected: Any?, actual: Any?
    ): Boolean = if (super.isExpected(expectedTable, actualTable, rowNum, columnName, dataType, expected, actual)) true
    else compareIgnoringMillis(expected, actual)

    private fun compareIgnoringMillis(expected: Any?, actual: Any?): Boolean {
        val expectedDt = LocalDateTime.fromDateFields(expected as Date).withMillisOfSecond(0)
        val actualDt = LocalDateTime.fromDateFields(actual as Timestamp)
        return expectedDt.isEqual(actualDt) || expectedDt.plusSeconds(1).isEqual(actualDt)
    }
}

private fun Any?.isNotNull() = this != null && this.toString().startsWith("!{notNull}")
private fun Any?.isNumber() = this != null && this.toString().startsWith("!{number}")
private fun Any?.isString() = this != null && this.toString().startsWith("!{string}")
private fun Any?.isRegex() = this != null && this.toString().startsWith("!{regex}")
private fun Any?.isWithin() = this != null && this.toString().startsWith("!{within ")

private fun String.withinPeriod() = parsePeriod(
    this.substring(
        "!{within ".length,
        this.indexOf("}")
    ).trim()
).toPeriod().toStandardDuration().millis

fun sortedTable(table: ITable, columns: Array<String>, rowComparator: RowComparator) = SortedTable(table, columns).apply {
    setUseComparable(true)
    setRowComparator(rowComparator.init(table, columns))
}